substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_sun_security_ssl_TrustStoreManager.java

```diff
@@ -94,27 +109,95 @@ public void afterRegistration(AfterRegistrationAccess access) {
 @TargetClass(className = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME)
 final class Target_sun_security_ssl_TrustStoreManager {
 
+    @Alias @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.NewInstance, declClassName = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME +
+                    "$TrustAnchorManager") private static Target_sun_security_ssl_TrustStoreManager_TrustAnchorManager tam = null;
+
     @Substitute
     private static Set<X509Certificate> getTrustedCerts() throws Exception {
-        return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedCerts;
-    }
+        Set<X509Certificate> set = new HashSet<>(ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedCerts);
+        set.addAll(tam.getTrustedCerts(Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor.createInstance()));
 
-    @Substitute
-    private static KeyStore getTrustedKeyStore() throws Exception {
-        return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedKeyStore;
+        return Collections.unmodifiableSet(set);
     }
 }
 
-/*
- * The internal classes to describe and load root certificates must not be reachable at run time.
- */
-
-@Delete
 @TargetClass(className = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME, innerClass = "TrustStoreDescriptor")
 final class Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor {
+
+    @Alias String storeName;
+    @Alias File storeFile;
+    @Alias String storeProvider;
+    @Alias String storeType;
+    @Alias String storePassword;
+
+    @Alias
+    static Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor createInstance() {
+        return null;
+    }
 }
 
-@Delete
 @TargetClass(className = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME, innerClass = "TrustAnchorManager")
 final class Target_sun_security_ssl_TrustStoreManager_TrustAnchorManager {
+
+    @SuppressWarnings({"unused", "static-method"}) //
+    @Alias
+    Set<X509Certificate> getTrustedCerts(Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor descriptor) throws Exception {
+        return null;
+    }
+
+    @Alias
+    private static KeyStore loadKeyStore(
+                    Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor descriptor) throws Exception {
+        KeyStore embeddedKeystore = ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedKeyStore;
+        if (!"NONE".equals(descriptor.storeName) &&
+                        descriptor.storeFile == null) {
+
+            // No file available, no KeyStore available.
+            if (SSLLogger.isOn && SSLLogger.isOn("trustmanager")) {
+                SSLLogger.fine("No available key store");
+            }
+
+            return embeddedKeystore;
```

Comment:
Without doing a full method diff, this seems to be the only change compared to the original implementation.
I don't really understand how that should work. The comment still says "no KeyStore available" but then something is returned?

Reply:
Yup that's the only difference...

You're right, maybe I should change the SSLLogger message to say that we found no keystore at runtime and we fall back to the embedded one.